cloud-nio/cloud-nio-impl-drs/src/main/scala/cloud/nio/impl/drs/DrsCloudNioFileProvider.scala

```diff
@@ -2,64 +2,107 @@ package cloud.nio.impl.drs
+      case Some(accessToken) if accessTokenTTLIsAcceptable(accessToken) => accessToken.getTokenValue
+      case _ =>
+        credential.refresh()
+        credential.getAccessToken.getTokenValue
     }
   }
 
 
+  private def checkIfPathExistsThroughMartha(drsPath: String): Boolean = {
+    val pathExistsIO = drsPathResolver.rawMarthaResponse(drsPath).use { marthaResponse =>
+      val errorMsg = s"Status line was null for martha response $marthaResponse."
+      DrsCloudNioFileProvider.convertOptionToIOWithErrorMsg(Option(marthaResponse.getStatusLine), errorMsg)
+    }.map(_.getStatusCode == HttpStatus.SC_OK)
+
+    pathExistsIO.unsafeRunSync()
+  }
+
+
   override def existsPath(cloudHost: String, cloudPath: String): Boolean =
     checkIfPathExistsThroughMartha(getDrsPath(cloudHost, cloudPath))
 
+
   override def existsPaths(cloudHost: String, cloudPathPrefix: String): Boolean =
     existsPath(cloudHost, cloudPathPrefix)
 
+
   override def listObjects(cloudHost: String, cloudPathPrefix: String, markerOption: Option[String]): CloudNioFileList = {
     val exists = existsPath(cloudHost, cloudPathPrefix)
     val list = if (exists) List(cloudPathPrefix) else Nil
     CloudNioFileList(list, None)
   }
 
+
   override def copy(sourceCloudHost: String, sourceCloudPath: String, targetCloudHost: String, targetCloudPath: String): Unit =
     throw new UnsupportedOperationException("DRS currently doesn't support copy.")
 
+
   override def deleteIfExists(cloudHost: String, cloudPath: String): Boolean =
     throw new UnsupportedOperationException("DRS currently doesn't support delete.")
 
-  override def read(cloudHost: String, cloudPath: String, offset: Long): ReadableByteChannel =
-    throw new UnsupportedOperationException("DRS currently doesn't support read.")
+
+  override def read(cloudHost: String, cloudPath: String, offset: Long): ReadableByteChannel = {
+    val drsPath = getDrsPath(cloudHost,cloudPath)
+    val freshAccessToken = getFreshAccessToken(authCredentials)
+
+    val byteChannelIO = for {
+      marthaResponse <- drsPathResolver.resolveDrsThroughMartha(drsPath, Option(freshAccessToken))
+      byteChannel <- drsReadInterpreter(marthaResponse)
+    } yield byteChannel
+
+    byteChannelIO.handleErrorWith {
+        e => IO.raiseError(new RuntimeException(s"Error while reading from DRS path: $drsPath. Error: ${ExceptionUtils.getMessage(e)}"))
+    }.unsafeRunSync()
+  }
+
 
   override def write(cloudHost: String, cloudPath: String): WritableByteChannel =
     throw new UnsupportedOperationException("DRS currently doesn't support write.")
 
+
   override def fileAttributes(cloudHost: String, cloudPath: String): Option[CloudNioRegularFileAttributes] =
     Option(new DrsCloudNioRegularFileAttributes(config, getDrsPath(cloudHost,cloudPath), drsPathResolver))
 }
+
+
+object DrsCloudNioFileProvider {
+
+  def convertOptionToIOWithErrorMsg[A](option: Option[A], errorMsg: String): IO[A] = {
```

Comment:
can this just be called `toIO` ?

Reply:
👍

Reply:
also would you please put this in cromwell common, perhaps in a package object for package `common.exception` ?

Reply:
Shifted `toIO` method to `common.exception` package object 👍